File: app/src/main/java/org/mifos/mobilebanking/models/Charge.java
package org.mifos.mobilebanking.models;

import android.os.Parcel;
import android.os.Parcelable;

import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

import org.mifos.mobilebanking.api.local.SelfServiceDatabase;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Vishwajeet
 * @since 16/8/16.
 */
@Table(database = SelfServiceDatabase.class)
public class Charge extends BaseModel implements Parcelable {

    @PrimaryKey
    Integer id;

    Integer clientId;
    Integer chargeId;

    @Column
    String name;

    List<Integer> dueDate = new ArrayList<>();
    ChargeTimeType chargeTimeType;
    ChargeCalculationType chargeCalculationType;
    Currency currency;

    @Column
    Double amount = 0.0;

    @Column
    Double amountPaid = 0.0;

    @Column
    Double amountWaived = 0.0;

    @Column
    Double amountWrittenOff = 0.0;

    @Column
    Double amountOutstanding = 0.0;

    Boolean penalty = false;

    @Column
    Boolean isActive = false;

    Boolean isPaid = false;
    Boolean isWaived = false;

    Boolean paid = false;
    Boolean waived = false;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getChargeId() {
        return chargeId;
    }

    public void setChargeId(int chargeId) {
        this.chargeId = chargeId;
    }

    public int getClientId() {
        return chargeId;
    }

    public void setClientId(int clientId) {
        this.clientId = clientId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public double getAmountPaid() {
        return amountPaid;
    }

    public void setAmountPaid(double amountPaid) {
        this.amountPaid = amountPaid;
    }

    public double getAmountWaived() {
        return amountWaived;
    }

    public void setAmountWaived(double amountWaived) {
        this.amountWaived = amountWaived;
    }

    public double getAmountWrittenOff() {
        return amountWrittenOff;
    }

    public void setAmountWrittenOff(double amountWrittenOff) {
        this.amountWrittenOff = amountWrittenOff;
    }

    public double getAmountOutstanding() {
        return amountOutstanding;
    }

    public void setAmountOutstanding(double amountOutstanding) {
        this.amountOutstanding = amountOutstanding;
    }

    public List<Integer> getDueDate() {
        return dueDate;
    }

    public void setDueDate(List<Integer> dueDate) {
        this.dueDate = dueDate;
    }

    public Currency getCurrency() {
        return currency;
    }

    public void setCurrency(Currency currency) {
        this.currency = currency;
    }

    public ChargeTimeType getChargeTimeType() {
        return chargeTimeType;
    }

    public void setChargeTimeType(ChargeTimeType chargeTimeType) {
        this.chargeTimeType = chargeTimeType;
    }

    public ChargeCalculationType getChargeCalculationType() {
        return chargeCalculationType;
    }

    public void setChargeCalculationType(ChargeCalculationType chargeCalculationType) {
        this.chargeCalculationType = chargeCalculationType;
    }

    public boolean isPenalty() {
        return penalty;
    }

    public void setPenalty(boolean penalty) {
        this.penalty = penalty;
    }

    public boolean isIsActive() {
        return isActive;
    }

    public void setActive(boolean isActive) {
        this.isActive = isActive;
    }

    public boolean isIsPaid() {
        return isPaid;
    }

    public void setPaid(boolean isPaid) {
        this.isPaid = isPaid;
    }

    public boolean isIsWaived() {
        return isWaived;
    }

    public void setWaived(boolean isWaived) {
        this.isWaived = isWaived;
    }

    public Boolean getPaid() {
        return paid;
    }

    public void setPaid(Boolean paid) {
        this.paid = paid;
    }

    public Boolean getWaived() {
        return waived;
    }

    public void setWaived(Boolean waived) {
        this.waived = waived;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.id);
        dest.writeValue(this.clientId);
        dest.writeValue(this.chargeId);
        dest.writeString(this.name);
        dest.writeList(this.dueDate);
        dest.writeParcelable(this.chargeTimeType, flags);
        dest.writeParcelable(this.chargeCalculationType, flags);
        dest.writeParcelable(this.currency, flags);
        dest.writeValue(this.amount);
        dest.writeValue(this.amountPaid);
        dest.writeValue(this.amountWaived);
        dest.writeValue(this.amountWrittenOff);
        dest.writeValue(this.amountOutstanding);
        dest.writeValue(this.penalty);
        dest.writeValue(this.isActive);
        dest.writeValue(this.isPaid);
        dest.writeValue(this.isWaived);
        dest.writeValue(this.paid);
        dest.writeValue(this.waived);
    }

    public Charge() {
    }

    protected Charge(Parcel in) {
        this.id = (Integer) in.readValue(Integer.class.getClassLoader());
        this.clientId = (Integer) in.readValue(Integer.class.getClassLoader());
        this.chargeId = (Integer) in.readValue(Integer.class.getClassLoader());
        this.name = in.readString();
        this.dueDate = new ArrayList<>();
        in.readList(this.dueDate, Integer.class.getClassLoader());
        this.chargeTimeType = in.readParcelable(ChargeTimeType.class.getClassLoader());
        this.chargeCalculationType = in.readParcelable(ChargeCalculationType.class.
                getClassLoader());
        this.currency = in.readParcelable(Currency.class.getClassLoader());
        this.amount = (Double) in.readValue(Double.class.getClassLoader());
        this.amountPaid = (Double) in.readValue(Double.class.getClassLoader());
        this.amountWaived = (Double) in.readValue(Double.class.getClassLoader());
        this.amountWrittenOff = (Double) in.readValue(Double.class.getClassLoader());
        this.amountOutstanding = (Double) in.readValue(Double.class.getClassLoader());
        this.penalty = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.isActive = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.isPaid = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.isWaived = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.paid = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.waived = (Boolean) in.readValue(Boolean.class.getClassLoader());
    }

    public static final Parcelable.Creator<Charge> CREATOR = new Parcelable.Creator<Charge>() {
        @Override
        public Charge createFromParcel(Parcel source) {
            return new Charge(source);
        }

        @Override
        public Charge[] newArray(int size) {
            return new Charge[size];
        }
    };
}

File: app/src/main/java/org/mifos/mobile/models/Charge.kt
package org.mifos.mobile.models

import android.os.Parcel
import android.os.Parcelable
import com.raizlabs.android.dbflow.annotation.Column
import com.raizlabs.android.dbflow.annotation.PrimaryKey
import com.raizlabs.android.dbflow.annotation.Table
import com.raizlabs.android.dbflow.structure.BaseModel
import org.mifos.mobile.api.local.SelfServiceDatabase

/**
 * @author Vishwajeet
 * @since 16/8/16.
 */
@Table(database = SelfServiceDatabase::class)
class Charge : BaseModel, Parcelable {
    @JvmField
    @PrimaryKey
    var id: Int? = null

    var clientId: Int? = null
    private var chargeId: Int? = null

    @Column
    var name: String? = null

    var dueDate: List<Int?> = ArrayList()
    private var chargeTimeType: ChargeTimeType? = null
    private var chargeCalculationType: ChargeCalculationType? = null
    var currency: Currency? = null

    @Column
    var amount = 0.0

    @Column
    var amountPaid = 0.0

    @Column
    var amountWaived = 0.0

    @Column
    var amountWrittenOff = 0.0

    @Column
    var amountOutstanding = 0.0

    var penalty = false

    @Column
    var isActive = false

    var isChargePaid: Boolean = false
    var isChargeWaived: Boolean = false

    var paid = false
    var waived = false

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(id)
        dest.writeValue(clientId)
        dest.writeValue(chargeId)
        dest.writeString(name)
        dest.writeList(dueDate)
        dest.writeParcelable(chargeTimeType, flags)
        dest.writeParcelable(chargeCalculationType, flags)
        dest.writeParcelable(currency, flags)
        dest.writeValue(amount)
        dest.writeValue(amountPaid)
        dest.writeValue(amountWaived)
        dest.writeValue(amountWrittenOff)
        dest.writeValue(amountOutstanding)
        dest.writeValue(penalty)
        dest.writeValue(isActive)
        dest.writeValue(isChargePaid)
        dest.writeValue(isChargeWaived)
        dest.writeValue(paid)
        dest.writeValue(waived)
    }

    constructor()
    private constructor(`in`: Parcel) {
        id = `in`.readValue(Int::class.java.classLoader) as Int
        clientId = `in`.readValue(Int::class.java.classLoader) as Int
        chargeId = `in`.readValue(Int::class.java.classLoader) as Int
        name = `in`.readString()
        dueDate = ArrayList()
        `in`.readList(dueDate, Int::class.java.classLoader)
        chargeTimeType = `in`.readParcelable(ChargeTimeType::class.java.classLoader)
        chargeCalculationType = `in`.readParcelable(ChargeCalculationType::class.java.classLoader)
        currency = `in`.readParcelable(Currency::class.java.classLoader)
        amount = `in`.readValue(Double::class.java.classLoader) as Double
        amountPaid = `in`.readValue(Double::class.java.classLoader) as Double
        amountWaived = `in`.readValue(Double::class.java.classLoader) as Double
        amountWrittenOff = `in`.readValue(Double::class.java.classLoader) as Double
        amountOutstanding = `in`.readValue(Double::class.java.classLoader) as Double
        penalty = `in`.readValue(Boolean::class.java.classLoader) as Boolean
        isActive = `in`.readValue(Boolean::class.java.classLoader) as Boolean
        isChargePaid = `in`.readValue(Boolean::class.java.classLoader) as Boolean
        isChargeWaived = `in`.readValue(Boolean::class.java.classLoader) as Boolean
        isChargePaid = `in`.readValue(Boolean::class.java.classLoader) as Boolean
        isChargeWaived = `in`.readValue(Boolean::class.java.classLoader) as Boolean
    }

    companion object {
        val CREATOR: Parcelable.Creator<Charge?> = object : Parcelable.Creator<Charge?> {
            override fun createFromParcel(source: Parcel): Charge {
                return Charge(source)
            }

            override fun newArray(size: Int): Array<Charge?> {
                return arrayOfNulls(size)
            }
        }
    }
}
